Monacs.Core/Either.cs

```diff
@@ -0,0 +1,147 @@
+            && (other.Left != null && other.Left.Equals(LeftValue)
+                || other.Right != null && other.Right.Equals(RightValue));
+
+        public override bool Equals(object obj) =>
+            !ReferenceEquals(null, obj)
+            && obj is IEither either && Equals(either);
+
+        public override int GetHashCode()
+        {
+            unchecked
+            {
+                var hashCode = EqualityComparer<L>.Default.GetHashCode(LeftValue);
+                hashCode = (hashCode * 397) ^ EqualityComparer<R>.Default.GetHashCode(RightValue);
+                hashCode = (hashCode * 397) ^ (int)LeftOrRight;
+                return hashCode;
+            }
+        }
+
+        public static bool operator ==(in Either<L, R> a, in Either<L, R> b) => a.Equals(b);
+
+        public static bool operator !=(in Either<L, R> a, in Either<L, R> b) => !a.Equals(b);
+
+        public override string ToString()
+        {
+            var side = IsLeft ? "Left" : "Right";
+            var value = IsLeft ? LeftValue.ToString() : RightValue.ToString();
+            return $"Either<{typeof(L).Name}, {typeof(R).Name}>({side} = {value})";
+        }
+    }
+
+    public enum LeftOrRight { Left, Right }
+
+    public static partial class Either
+    {
+        /* Match */
+
+        // E<L,R> -> Out
+        public static Out Match<L, R, Out>(this Either<L, R> either, Func<L, Out> left, Func<R, Out> right) =>
+            either.IsLeft ? left(either.LeftValue) : right(either.RightValue);
+
+        /* Map */
+
+        // E<L,R> -> (L -> LOut, R -> ROut) -> E<LOut, ROut>
+        public static Either<LOut, ROut> Map<L, R, LOut, ROut>(this Either<L, R> either, Func<L, LOut> mapLeft, Func<R, ROut> mapRight) =>
+            either.IsLeft ? ToEitherLeft<LOut, ROut>(mapLeft(either.LeftValue)) : ToEitherRight<LOut, ROut>(mapRight(either.RightValue));
+
+        /* Bind */
+
+        // E<L, R> -> (L -> LOut, R -> ROut) -> E<LOut, ROut>
+        public static Either<LOut, ROut> Bind<L, R, LOut, ROut>(this Either<L, R> either, Func<L, Either<LOut, ROut>> leftBinder, Func<R, Either<LOut, ROut>> rightBinder) =>
+            either.IsLeft ? leftBinder(either.LeftValue) : rightBinder(either.RightValue);
+
+
+        /* Side effects */
+
+        // E<L, R> -> (L -> Unit) -> E<L, R>
+        public static Either<L, R> DoWhenLeft<L, R>(this Either<L, R> either, Action<L> action)
+        {
+            if (either.IsLeft)
+                action(either.LeftValue);
+
+            return either;
+        }
+
+        // E<L, R> -> (R -> Unit) -> E<L, R>
+        public static Either<L, R> DoWhenRight<L, R>(this Either<L, R> either, Action<R> action)
+        {
+            if (either.IsRight)
+                action(either.RightValue);
+
+            return either;
+        }
+
+        /* Factories */
+
+        // L -> E<L, R>
+        public static Either<L, R> ToEitherLeft<L, R>(L left) => new Either<L, R>(left);
```

Comment:
This should be just called `Left`

Reply:
I was thinking about making it extension method. That way in C# we'd write `42.Left()` which looks a bit odd and `42.ToEitherLeft()` is looks much better.

Should we still stay with `Left`?

Reply:
Yes, `Left` is the best version IMO, but not an extension method - we don't use it in other types and it would make the API less consistent.